### PR TITLE
refactor: test suite

### PR DESCRIPTION
**Pull request by ryanio**
This PR addresses #125 and #132 to refactor and modernize the test suite.

1. Breaks up the single large `tape` test containing leaky shared state into individually siloed tests.
1. Updates the test fixture to one with metadata located [here](https://github.com/ethereum/tests/blob/develop/BlockchainTests/randomStatetest391.json).

Feedback and comments welcome! I am hoping this is a huge improvement.

### CODE REVIEW

**Comment by holgerd77** [comment]
Hi @ryanio, this is so so great, thanks 😊! This updates actually makes the whole library so much more approachable, just realized this for myself while scanning through the changes, since I didn't dig too deep into the `blockchain` library and always had my problems to approach and experiment with the code. Within a holistic view I would also put stuff like this as a lower level part of a developer improvement initiative, I think for many of the more experienced devs this will make the library more approachable, really great! It also better prepares the library for future ongoing development and extension, if we e.g. extend the library to also support chains from the Eth 2.0 suite. We'll see. 😄

Some first note: we currently seem to have a general **problem with coverage**, see this issue https://github.com/ethereumjs/ethereumjs-config/issues/22. Beyond coverage is not working with these changes from the PR since the base config from the `config` library is set to work on the `src` folder. Think we need to give this some separate thinking, I just wanted to have the coverage results, so I ran this manually by runnning:

```shell
./node_modules/nyc/bin/nyc.js npm run test
```

and modifying the `.nycrc` file to:

```json
{
  "extends": "@ethereumjs/config-nyc",
  "include": [
    "dist/**/*.js"
  ]
}
```

Bildschirmfoto 2019-12-02 um 12 06 46

This might be a slight drop - not sure yet, the [latest reported number](https://coveralls.io/github/ethereumjs/ethereumjs-blockchain) here was on 96%. I nevertheless think we can live with this regarding the scope of the changes here. Latest coverage report is also quite old (from February 2019) so this might very well due to new code additions and not due to the changes here in the test suite.

So again:

👍 😄

File: test/index.ts [hunk 1]
```diff
@@ -1,805 +1,1080 @@
...
-    ],
-    (err?: Error) => {
-      cb(err, db, genesis)
-    },
-  )
-}
+      st.end,
+    )
+  })
+})
```
Comment (holgerd77) [inline test/index.ts]: Serially run tests broken down into separate `t.test` calls, nice.

File: test/util.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        key: 'heads',
+        valueEncoding: 'json',
+        value: { head0: { type: 'Buffer', data: [171, 205] } },
+      },
+    ],
+    (err?: Error) => {
+      cb(err, db, genesis)
+    },
+  )
+}
```
Comment (holgerd77) [inline test/util.ts]: Also good to have this stuff extracted.

File: test/index.ts [hunk 3]
```diff
@@ -1,805 +1,1080 @@
...
+    blockchain.putGenesis(genesisBlock, (err?: Error) => {
+      st.error(err, 'no error')
+      blockchain.putBlocks(blocks.slice(1), (err?: Error) => {
+        if (err) {
+          return st.error(err)
+        }
+        // start: 5, max: 5, skip: 1, reverse: true
+        blockchain.getBlocks(5, 5, 1, true, (err?: any, getBlocks?: any) => {
+          st.error(err, 'no error')
+          st.equal(getBlocks.length, 3)
```
Comment (s1na) [inline test/index.ts]: It should equal 5 after #137 is fixed. Or `getBlocks(10, 10, 1, ...)`, in which case the length should be 4 (I think)

File: test/index.ts [hunk 4]
```diff
@@ -1,805 +1,1080 @@
...
+            if (block.hash().equals(blocks[i + 1].hash())) {
+              i++
+            }
             cb()
           },
-          function() {
-            t.equals(i, 9, 'should iterate through 9 blocks')
-            done()
+          () => {
+            st.equals(i, 15)
```
Comment (s1na) [inline test/index.ts]: Not sure, but I feel like this should be 14?

Update: nvm realized `generateBlocks` generates n blocks in addition to the genesis blocks.

File: test/index.ts [hunk 5]
```diff
@@ -1,805 +1,1080 @@
 import * as async from 'async'
 import Common from 'ethereumjs-common'
-import { rlp, toBuffer } from 'ethereumjs-util'
+import { toBuffer } from 'ethereumjs-util'
 import * as test from 'tape'
-import Blockchain from '../src'
+import Blockchain from '../dist'
```
[the remark this answers is not in the document]
Comment (ryanio) [reply test/index.ts]: I thought it would make sense to run the tests against the built version, since that would be what people would be receiving via package managers.

File: test/index.ts [hunk 6]
```diff
@@ -1,805 +1,1080 @@
...
-        t.throws(
-          function() {
-            new Blockchain({ chain: 'ropsten', common: common })
-          },
-          /not allowed!$/,
-          'should throw on initialization with chain and common parameter',
-        )
+test('blockchain test', t => {
+  t.test('should not crash on getting head of a blockchain without a genesis', st => {
+    const blockchain = new Blockchain({ validate: false })
```
[the remark this answers is not in the document]
Comment (ryanio) [reply test/index.ts]: gotcha, will do, thank you

[the remark this answers is not in the document]
Comment (ryanio) [reply test/index.ts]: I am getting `invalid difficulty` with `validateBlocks: true` when setting difficulty to `'0xfffffff'`. Is there a value I can set it to that will pass validation, or should I set `validateBlocks: false` for these tests?

**Comment by holgerd77** [comment]
Hi @s1na, thanks for rescuing me here, realizing that my looking-over was too superficial. 🤨

**Comment by ryanio** [comment]
Thanks for your comments @s1na!

I am running into a weird edge case that a newly initialized `Blockchain` returns `1` block. I added a failing test case. Any idea what could be going on here?

I tried to grasp your comments on the `getBlock` tests and integrated them. They might still need some adjustment, and maybe some better comments inline to describe what we are testing.

File: test/index.ts (hunk 4 above)
Comment (ryanio) [reply test/index.ts]: I have changed this a bit now with the new `generateBlockchain` util, should be easier to understand.

**Comment by holgerd77** [comment]
@ryanio Can you update the branch here respectively rebase?

File: test/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (s1na) [reply test/index.ts]: Hm I don't know what the best practice here is, but your reasoning makes sense to me, so let's keep it `dist` unless someone else objects

File: test/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (s1na) [reply test/index.ts]: Hm :thinking: can you try setting the difficulty (after creating the block) to the canonical difficulty (via this [method](https://github.com/ethereumjs/ethereumjs-block/blob/15dc2a70aae0856f78cae35a14ab4b65d9537ff3/src/header.ts#L134))? 
It'd be nice to also to have `validateBlocks` in the tests, but if it was difficult to get right for the mocked blocks I guess we can use `false`.

**Comment by s1na** [comment]
You're right Ryan, it seems that `Blockchain` is purposefully creating a genesis block on initialization if one doesn't already exist:

https://github.com/ethereumjs/ethereumjs-blockchain/blob/5edfd80e39c22384339e3c3c7a170a10e99fa0d0/src/index.ts#L264-L265

It might be a bit confusing but I don't see a harm in it. Still we can consider changing it if there's consensus

**Comment by holgerd77** [comment]
> You're right Ryan, it seems that `Blockchain` is purposefully creating a genesis block on initialization if one doesn't already exist:
> 
> https://github.com/ethereumjs/ethereumjs-blockchain/blob/5edfd80e39c22384339e3c3c7a170a10e99fa0d0/src/index.ts#L264-L265
> 
> It might be a bit confusing but I don't see a harm in it. Still we can consider changing it if there's consensus

Maybe in a first iteration this can be just a bit better documented?

**Comment by ryanio** [comment]
> Maybe in a first iteration this can be just a bit better documented?

Thanks @s1na for investigating!

Ok I'm thinking I'll update this PR to just refactor the test suite as the library is currently written and add some additional clarification in the documentation. I'll open a separate PR for removing skip if we don't find anyone using it.

File: test/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ryanio) [reply test/index.ts]: Great, thanks for that! Got it working so switched all validateBlocks to `true` :)

**Comment by ryanio** [comment]
Ok! Should be good to go now :)

@s1na I would appreciate a final check that the tests look good.

**Comment by alcuadrado** [comment]
Thanks a lot for this change @ryanio! I didn't have enough time to review it, but it's a great improvement. I was always scared of changing these tests.
